### DISCUSSION TITLE
Enum "name" instead of "value" in route def (and docs)

### DISCUSSION THREAD

**Discussion by WilliamStam** [fastapi/fastapi#11098] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import enum
from fastapi import FastAPI

app = FastAPI()

class MyEnum(str, enum.Enum):
    WOOF = "a"
    MEOW = "b"

from typing import List, Literal

@app.get("")
async def testing_enums(
    myenum: MyEnum = None,
) -> None:
   print(myenum)
```

### Description

this is sort of a pydantic thing since pydantic handles the validation, but i would assume that fastapi handles the "allowed" options in the openapi spec

when use'ing the above code the "allowed" options are "a" and "b", so defining an enum with key and value the same is fine but as soon as you want to use any other value for the enums but select by name then things get _interesting_. this naturally also invalidates the use of `enum.auto()` 

how would i go about having fastapi give the options the user can enter as `WOOF` and `MEOW` instead of `a` and `b` (or in the case of enum.auto() its ints `1` or `2`

### Operating System

Windows

### FastAPI Version

0.109.0

### Pydantic Version

2.6.1

### Python Version

3.12.1

**Comment by msehnout** [comment]
Just rewriting the enum like this does not solve the issue?
```python
class MyEnum(str, enum.Enum):
    WOOF = "WOOF"
    MEOW = "MEOW"
```

**Reply by WilliamStam** [reply]
nope. consider pythons enum.auto() outputs an int yet the name is a str.

```
class MyEnum(enum.Enum):
   dog = enum.auto()
   cat = enum.auto()

```

shows as "allowed 1 or 2"

my use case is a lot more complex than that tho but i think it gets the point across 

----

actual implementation is to the lines of

```
class order(BaseOrdering):
    name = OrderColumn(UsersTable.name)
    username = OrderColumn(UsersTable.username)


@app.get("")
async def testing_enums(
    order: order= None, # it should say "allowed 'name' or 'username'"
)

```

**Reply by JavierSanchezCastro** [reply]
You can check these two links:

https://stackoverflow.com/questions/67911340/encode-pydantic-field-using-the-enum-name-instead-of-the-value
https://fastapi-utils.davidmontague.xyz/user-guide/basics/enums/

Although @msehnout gives you the correct and easy solution for that

**Comment by WilliamStam** [comment, marked as answer]
@JavierSanchezCastro send me down a bit of a rabbit hole but in a strange messed up way it got me to an answer so thanks! 

using @msehnout 's answer wasn't the answer.

so in the end i went with hijacking the enum object,to actually return a name: name and on each member to set an orig object of what that value was originally

here's the final form i came up with:


```
# utilities/enum.py

from enum import Enum
class NameEnum(Enum):
    """
    FastAPI needs the "value" to be the lookup value.
    Using this class we can "fool" fastapi to believe the enum is

    ```
    class MyEnum(Enum):
        MEMBER1 = "MEMBER1"
        MEMBER2 = "MEMBER2"
    ```

    Even tho it's something completely different. the original value is set behind the .orig property

    Usage:

    class MyEnum(NameEnum):
        MEMBER1 = 99
        MEMBER2 = 69

    assert MyEnum.MEMBER1.value == "MEMBER1" # True
    assert MyEnum.MEMBER1.value == 99 # False

    assert MyEnum.MEMBER1.orig == 99 # True
    assert MyEnum.MEMBER1.orig == "MEMBER1 # False

    Since the "value" is immaterial you can use whatever value you want including enum.auto().

    """

    def __init__(self, *args):
        self._old = self._value_
        self._value_ = self._name_

    @classmethod
    def _missing_(cls, value):
        for member in cls:
            if member.name == value:
                return member
        return None

    @property
    def orig(self):
        return self._old

```

```
# repository.py



class order_by(NameEnum): # the values aren't important
    name = OrderColumn(UsersTable.name) 
    username = OrderColumn(UsersTable.username)

async def find(
    order: order_by | None = None,
) -> PaginatedCollection[model.UserModel]:

    logger.info(order)
```


```
# router.py

async def get_all_users(
    order: repository.order_by = repository.order_by.name,
) -> UsersResponse:

    response = UsersResponse()
    response.users = repository.find(order=order)
```

and it shows up nicely in the final docs

![image](https://github.com/tiangolo/fastapi/assets/182800/301b8d7b-5e02-4bd6-9181-f10ea03e567d)

**Comment by comfuture** [comment]
I found another solution from another reference. Instead of replacing `._value_`, use generics with a custom serializer. OpenAPI Spec also works fine.

```python
from typing import Annotated, Type, TypeVar
from enum import Enum
from pydantic_core import CoreSchema, core_schema
from pydantic import (
    GetCoreSchemaHandler,
    GetJsonSchemaHandler,
    BaseModel,
)
from pydantic.json_schema import JsonSchemaValue


class EnumNameSerializer:
    """
    A serializer that produces `.name` instead of `.value`
    """

    def __get_pydantic_core_schema__(
        self, enum_cls: Type[Enum], _handler: GetCoreSchemaHandler
    ) -> CoreSchema:
        def customize_json_schema(
            _core_schema: CoreSchema, _handler: GetJsonSchemaHandler
        ) -> JsonSchemaValue:
            return {
                "type": "string",
                "enum": [e.name for e in enum_cls],
            }
        def enum_or_name(x: Enum | str) -> Enum:
            if isinstance(x, str):
                return enum_cls[x]
            return x

        s = core_schema.no_info_after_validator_function(
            enum_or_name,
            core_schema.str_schema(),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda e: e.name
            ),
        )
        s.setdefault("metadata", {}).setdefault("pydantic_js_functions", []).append(
            customize_json_schema
        )
        return s

EnumType = TypeVar("EnumType", bound=Enum)
NameEnum = Annotated[EnumType, EnumNameSerializer()]
```

You can use it like this:

```python
class Animal:
    cat: enum.auto()
    dog: enum.auto()
    monkey: enum.auto()

# or in shorthand
# Animal = IntEnum('Animal', 'cat dog monkey')

@app.route('/make_sound/{animal}')
async def make_sound(animal: NameEnum[Animal]):
    if animal == Animal.cat:
        return 'meow~'
    elif animal == Animal.dog:
        return 'bark!'
    elif animal == Animal.monkey:
        return 'ukiki'
```


https://github.com/fastapi/fastapi/discussions/12003

I didn't find this topic when I was writing another discussion topic. Nice to meet people who have the same concerns.

**Reply by comfuture** [reply]
Please Note that. If you specify a `Path` as a parameter default value, fastapi's `json_serialzable()` will revert it back to a value.

```python
@app.route('/make_sound/{animal}')
async def make_sound(animal: NameEnum[Animal] = Path(..., description='The animal'): # Bad
    # ...
```
